src/main/java/com/aws/greengrass/deployment/DefaultDeploymentTask.java

```diff
@@ -173,25 +179,45 @@ public DeploymentResult call() throws InterruptedException {
         }
     }
 
-
+    @SuppressWarnings("PMD.AvoidCatchingGenericException")
     private Map<String, Set<ComponentIdentifier>> getNonTargetGroupToRootPackagesMap(
             DeploymentDocument deploymentDocument)
             throws DeploymentTaskFailureException, InterruptedException {
-        Map<String, Set<ComponentIdentifier>> nonTargetGroupsToRootPackagesMap = new HashMap<>();
 
+        // Don't block local deployments due to device being offline by using finite retries for getting the
+        // hierarchy and fall back to hierarchy stored previously in worst case. For cloud deployment, use infinite
+        // retries by default similar/to all other cloud interactions.
+        boolean isLocalDeployment = Deployment.DeploymentType.LOCAL.equals(deployment.getDeploymentType());
+        int retryCount = isLocalDeployment ? FINITE_RETRY_COUNT : INFINITE_RETRY_COUNT;
+
+        Optional<Set<String>> groupsForDeviceOpt = Optional.empty();
+        AtomicBoolean useLocalMapping = new AtomicBoolean();
+        try {
+            groupsForDeviceOpt = thingGroupHelper.listThingGroupsForDevice(retryCount);
+        } catch (Exception e) {
+            if (isLocalDeployment && ThingGroupHelper.DEVICE_OFFLINE_INDICATIVE_EXCEPTIONS.contains(e.getClass())) {
+                logger.atWarn().setCause(e).log("Failed to get thing group hierarchy, local deployment will proceed");
+                useLocalMapping.set(true);
```

Comment:
Do we need this ? setting groupsForDeviceOpt as Optional.empty() should have the desired effect right?

Reply:
That will end up removing the device from all groups